allennlp/data/dataset_readers/dataset_reader.py

```diff
@@ -334,77 +388,66 @@ def _instance_iterator(self, file_path: str) -> Iterable[Instance]:
+            if self.max_instances is not None:
+                # But we don't write to the cache when max_instances is specified.
+                logger.warning("Skipping writing to data cache since max_instances was specified.")
+                yield from instances
+            elif util.is_distributed() or (get_worker_info() and get_worker_info().num_workers):
+                # We also shouldn't write to the cache if there's more than one process loading
                 # instances since each worker only receives a partial share of the instances.
                 logger.warning(
                     "Can't cache data instances when there are multiple processes loading data"
                 )
-                # Release the lock, then just read the instances as normal.
-                cache_file_lock.release()
-                for instance in self._multi_worker_islice(self._read(file_path)):
-                    yield instance
+                yield from instances
             else:
-                # Otherwise we can safely write to it.
                 try:
-                    with open(cache_file, "w") as data_file:
-                        for instance in self._multi_worker_islice(self._read(file_path)):
-                            data_file.write(self.serialize_instance(instance) + "\n")
-                            yield instance
-                except:  # noqa: E722
-                    # If anything went wrong, the cache file will be corrupted, so we should
-                    # remove it.
-                    if os.path.exists(cache_file):
-                        logger.warning(
-                            "Removing dataset cache file '%s' due to exception", cache_file
-                        )
-                        os.remove(cache_file)
-                    raise
-                finally:
-                    # Release the lock no matter what.
-                    cache_file_lock.release()
+                    with FileLock(cache_file + ".lock", timeout=self.CACHE_FILE_LOCK_TIMEOUT):
+                        with tempfile.NamedTemporaryFile("w+") as temp_file:
+                            logger.info("Caching instances to temp file %s", temp_file.name)
+                            for instance in instances:
+                                temp_file.write(self.serialize_instance(instance) + "\n")
+                                yield instance
+
+                            # we are copying the file before closing it, so flush to avoid truncation
+                            temp_file.flush()
+                            # shutil.copyfileobj() starts at the current position, so go to the start
+                            temp_file.seek(0)
+
+                            logger.info("copying %s to cache at %s", temp_file.name, cache_file)
+                            with open(cache_file, "w") as cache:
+                                shutil.copyfileobj(temp_file, cache)
```

Comment:
Is this here twice?

Reply:
This handles lazy caching, the other place is non-lazy caching. I thought about trying to combine the logic but it didn't work out.